### PR TITLE
Limit the number of records returned as per the page_size defined

### PR DESCRIPTION
**Pull request by bhuvana-talend**
# Description of change
https://jira.talendforge.org/browse/TDL-16124
client.py:

Modified read to accept page_size and pagination=True/False as option parameter
Modified Get methods to pass these parameters wherever pagination is controlled by the page_size

test_hubspot_all_fields.py
test_hubspot_pagination.py
test_hubspot_start_date.py

Modified these tests to pass page_size in the call to get methods and read methods
# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/client.py [hunk 1]
```diff
@@ -322,6 +325,8 @@ def get_contact_lists(self, since='', list_id=''):
 
             has_more = response['has-more']
             params['offset'] = response['offset']
+            if pagination and len(records) > 260:
```
[the remark this answers is not in the document]
Comment (bhuvana-talend) [reply tests/client.py:328]: For contact_lists, this method is also called from another place where I have a problem passing the page size. We can discuss looking at the flow.

File: tests/client.py [hunk 2]
```diff
@@ -379,18 +384,22 @@ def get_contacts(self):
...
+
             has_more = response_1['has-more']
             params_1['vidOffset'] = response_1['vid-offset']
 
             # get the detailed contacts records by vids
             params_2['vid'] = vids
             response_2 = self.get(url_2, params=params_2)
             records.extend([record for record in response_2.values()])
+            if pagination  and len(records) > page_size+10:
+                LOGGER.info("haven't met the page size")
```
[the remark this answers is not in the document]
Comment (bhuvana-talend) [reply tests/client.py:396]: Yeah.. tis is for debug. Removed it

File: tests/client.py [hunk 3]
```diff
@@ -880,7 +905,7 @@ def create_contacts_by_company(self, company_ids=[], contact_records=[], times=1
         if not company_ids:
             company_ids = [company['companyId'] for company in self.get_companies()]
         if not contact_records:
-            contact_records = self.get_contacts()
+            contact_records = self.get_contacts(100)
```
[the remark this answers is not in the document]
Comment (bhuvana-talend) [reply tests/client.py:908]: We get page_size from expected_metadata in base.py
This is at create_contacts_by_company, where we don't have the ability to pass page_size. I started modifying for pagination test with the page_size and saw lot of hard coded values and tried to remove hard coded values as much as possible. But, it is not possible in a couple of places.

[the remark this answers is not in the document]
Comment (bhtowles) [reply tests/client.py:907]: Ok, I would like to tag all the places we have hard coded values and then follow up with Harrison or dev to see if there is any way around it that isn't more work than it's worth.  That way we have all of the places already marked if we ever need to update.

**Review by bhtowles — COMMENTED** [review]
Looks like build_daily is passing but this commit is not.  Perhaps try to merge main into this branch with the next commit to get tests passing.

File: tests/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhuvana-talend) [reply tests/client.py:909]: I have added a comment with TODO and Jira ticket wherever it is hardcoded. Will post it  in slack channel as a follow-up to get Dev's attention on this once this PR is merged.

**Comment by bhuvana-talend** [comment]
> Looks like build_daily is passing but this commit is not. Perhaps try to merge main into this branch with the next commit to get tests passing.

Tests are passing now

File: tests/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhuvana-talend) [reply tests/client.py]: @bhtowles , Made the changes as per our discussion yesterday. So, now, all changes are in client.py with a little change in test_hubspot_pagination.py

File: tests/client.py [hunk 4]
```diff
@@ -238,10 +239,11 @@ def _get_company_by_id(self, company_id):
         response = self.get(url)
         return response
 
-    def get_companies(self, since=''):
+    def get_companies(self, since='', pagination=False):
         """
         Get all companies by paginating using 'hasMore' and 'offset'.
         """
+        page_size = self.BaseTest.expected_metadata()['companies'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:246]:
```suggestion
        page_size = self.BaseTest.expected_metadata().get('companies', {}).get(self.BaseTest.EXPECTED_PAGE_SIZE)
```

File: tests/client.py [hunk 5]
```diff
@@ -283,11 +287,12 @@ def get_companies(self, since=''):
 
         return records
 
-    def get_contact_lists(self, since='', list_id=''):
+    def get_contact_lists(self, since='', list_id='', pagination=False):
         """
         Get all contact_lists by paginating using 'has-more' and 'offset'.
         """
         url = f"{BASE_URL}/contacts/v1/lists"
+        page_size = self.BaseTest.expected_metadata()['contact_lists'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:295]: same as above, we should avoid using dict keys directly in favor of get() with default fallback if there is more than one get()

File: tests/client.py [hunk 6]
```diff
@@ -354,16 +361,17 @@ def _get_contacts_by_pks(self, pks):
 
         return records[0]
 
-    def get_contacts(self):
+    def get_contacts(self, pagination=False):
         """
         Get all contact vids by paginating using 'has-more' and 'vid-offset/vidOffset'.
         Then use the vids to grab the detailed contacts records.
         """
+        page_size = self.BaseTest.expected_metadata().get('contacts').get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:369]:
```suggestion
        page_size = self.BaseTest.expected_metadata().get('contacts', {}).get(self.BaseTest.EXPECTED_PAGE_SIZE)
```

File: tests/client.py [hunk 7]
```diff
@@ -400,8 +411,9 @@ def get_contacts_by_company(self, parent_ids):
             pulling the 'companyId' from each record to perform the corresponding get here.
         """
 
+        page_size = self.BaseTest.expected_metadata()['contacts_by_company'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:414]: same comment

File: tests/client.py [hunk 8]
```diff
@@ -512,13 +526,14 @@ def get_deals(self):
         records = self.denest_properties('deals', records)
         return records
 
-    def get_email_events(self, recipient=''):
+    def get_email_events(self, recipient='', pagination=False):
         """
         Get all email_events by paginating using 'hasMore' and 'offset'.
         """
+        page_size = self.BaseTest.expected_metadata()['email_events'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:533]: same

File: tests/client.py [hunk 9]
```diff
@@ -549,13 +566,14 @@ def _get_engagements_by_pk(self, engagement_id):
 
         return response
 
-    def get_engagements(self):
+    def get_engagements(self, pagination=False):
         """
         Get all engagements by paginating using 'hasMore' and 'offset'.
         """
+        page_size = self.BaseTest.expected_metadata()['engagements'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:573]: same

File: tests/client.py [hunk 10]
```diff
@@ -606,13 +626,14 @@ def get_owners(self):
         transformed_records = self.datatype_transformations('owners', records)
         return transformed_records
 
-    def get_subscription_changes(self, since=''):
+    def get_subscription_changes(self, since='', pagination=False):
         """
         Get all subscription_changes from 'since' date by paginating using 'hasMore' and 'offset'.
         Default since date is one week ago
         """
+        page_size = self.BaseTest.expected_metadata()['subscription_changes'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:634]: same

File: tests/client.py [hunk 11]
```diff
@@ -677,18 +700,19 @@ def get_tickets_properties(self):
 
         return ",".join([record["name"] for record in records["results"]])
 
-    def get_tickets(self):
+    def get_tickets(self, pagination=False):
         """
         Get all tickets.
         HubSpot API https://developers.hubspot.com/docs/api/crm/tickets
         """
+        page_size = self.BaseTest.expected_metadata()['tickets'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:708]: same

File: tests/client.py [hunk 12]
```diff
@@ -880,7 +912,8 @@ def create_contacts_by_company(self, company_ids=[], contact_records=[], times=1
         if not company_ids:
             company_ids = [company['companyId'] for company in self.get_companies()]
         if not contact_records:
-            contact_records = self.get_contacts()
+            page_size = self.BaseTest.expected_metadata()['contacts_by_company'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:915]: same

File: tests/client.py [hunk 13]
```diff
@@ -1041,7 +1074,8 @@ def create_engagements(self):
         record_uuid = str(uuid.uuid4()).replace('-', '')
 
         # gather all contacts and randomly choose one that has not hit the limit
-        contact_records = self.get_contacts()
+        page_size = self.BaseTest.expected_metadata()['engagements'].get(self.BaseTest.EXPECTED_PAGE_SIZE)
```
Comment (bhtowles) [inline tests/client.py:1077]: same

**Review by bhtowles — APPROVED** [review]

**Pull request merged**